drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1516 @@
+#endif
+	ret = cdns_wait_sdma(base_address);
+	if (ret != 0) {
+		return ret;
+	}
+	sdma_buffer_copy(params->sdma_base, data_buffer, params->page_size, CDNS_READ);
+	return 0;
+}
+
+/**
+ * Perform a generic page write operation in the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param write_address The address to which the page will be written.
+ * @param data_buffer The buffer containing the data to be written.
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_generic_page_write(struct cadence_nand_params *params, uint64_t write_address,
+				   void *data_buffer)
+{
+	uint64_t mini_ctrl_cmd = 0;
+	int ret;
+
+	uintptr_t base_address = params->nand_base;
+
+	mini_ctrl_cmd |= GCMD_TWB_VALUE;
+	mini_ctrl_cmd |= GEN_ADDR_WRITE_DATA((uint32_t)write_address);
+	if ((params->nluns > 1) || (params->total_bit_row > BIT16_CHECK)) {
+		mini_ctrl_cmd |= PAGE_MAX_BYTES(PAGE_MAX_SIZE);
+	} else {
+		mini_ctrl_cmd |= PAGE_MAX_BYTES(PAGE_MAX_SIZE - 1);
+	}
+	mini_ctrl_cmd |= PAGE_WRITE_CMD;
+	ret = cdns_generic_send_cmd(params, mini_ctrl_cmd);
+	if (ret != 0) {
+		return ret;
+	}
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	k_sem_take(&params->interrupt_sem_t, K_FOREVER);
+#endif
+	ret = cdns_generic_cmd_data(params, CDNS_WRITE, params->page_size);
+	if (ret != 0) {
+		return ret;
+	}
+	sdma_buffer_copy(params->sdma_base, data_buffer, params->page_size, CDNS_WRITE);
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	k_sem_take(&params->interrupt_sem_t, K_FOREVER);
+#endif
+	mini_ctrl_cmd = 0;
+	mini_ctrl_cmd |= PAGE_WRITE_10H_CMD;
+	mini_ctrl_cmd |= GCMD_TWB_VALUE;
+	mini_ctrl_cmd |= PAGE_CMOD_CMD;
+	ret = cdns_generic_send_cmd(params, mini_ctrl_cmd);
+	if (ret != 0) {
+		return ret;
+	}
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	k_sem_take(&params->interrupt_sem_t, K_FOREVER);
+#endif
+	ret = cdns_wait_sdma(base_address);
+	return ret;
+}
+
+/**
+ * Perform a generic read or write operation for a range of pages in the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param start_page_number The starting page number for the read or write operation.
+ * @param buffer The buffer containing the data to be written or to store the read data.
+ * @param page_count The number of pages to be read or written.
+ * @param mode The mode of operation (read, write).
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_nand_gen_read_write(struct cadence_nand_params *params, uint32_t start_page_number,
+				    char *buffer, uint32_t page_count, uint8_t mode)
+{
+	uint64_t address = 0;
+	uint32_t index = 0;
+	uint32_t gen_row_address = 0;
+	int ret = 0;
+
+	for (index = 0; index < page_count; index++) {
+		row_address_set(params, &gen_row_address, start_page_number++);
+		address = ((uint64_t)gen_row_address);
+		if (mode == CDNS_READ) {
+			ret = cdns_generic_page_read(params, address,
```

Comment:
I guess when `ret` is error you keep on looping?

Reply:
added return condition if error occurs.